state/protocol/protocol_state/epochs/protocol_statemachine.go

```diff
@@ -26,10 +26,10 @@ type protocolStateMachine struct {
 
 var _ ProtocolStateMachine = (*protocolStateMachine)(nil)
 
-// newStateMachine creates a new protocol state protocolStateMachine.
+// NewStateMachine creates a new protocol state protocolStateMachine.
```

Comment:
I would strongly advocate for `parallel` state machines. If we want to avoid getting ourselves in a hell of logical interdependencies, I think there is only one self-consistent approach:
* Each state machine has its own portion of the state that it _exclusively_ mutates. 
* Therefore, the overall state is a product state of all the state-machine specific sub-states
* The overall state machine is formalized as `Orthogonal Regions` [[1](https://en.wikipedia.org/wiki/UML_state_machine#Orthogonal_regions), [2](https://www.state-machine.com/doc/Pattern_Orthogonal.pdf)] composed of the individual state machines. 

For an intuition, this is often framed as "concurrent state machines" (or product-state-machines). 

Note: 
* This is an explicit limitation, that we purposefully impose on the Protocol State's KV-store. Mathematically, we say that the overall state machine $\mathcal{S}$ must be decomposable into $\mathcal{S} = \mathcal{S}_0 \otimes \mathcal{S}_1 \otimes \ldots  \otimes \mathcal{S}_j$
* This does not prevent us of from implementing sequential state machines for certain events. It is just that we would need to bundle these sequential state machines into a composite state machine (conceptually a processing pipeline). Something like saying that $\mathcal{S}_i$ (for  some specific $i \in [0,\ldots,j]$) can be decomposed into  $\mathcal{S}_i = \mathcal{S}^{(\kappa)}_i \cdots \mathcal{S}^{(1)}_i\mathcal{S}^{(0)}_i$

I hope the mathematical notation helps to explain 😅 ($\otimes$ is not quite a [tensor product](https://en.wikipedia.org/wiki/Tensor_product) because we are not operating on vector spaces, but conceptually it is the same thing, product spaces and product operators)

Reply:
I think what we are running into here in this particular case is the lack of separation between:
* the framework that evolves the protocol state in my mind is [`MutableProtocolState`](https://github.com/onflow/flow-go/blob/ba1f763f08d56fff590b4c9c3e70399c72749f45/state/protocol/protocol_state/state/protocol_state.go#L83) and [`stateMutator`](https://github.com/onflow/flow-go/blob/ba1f763f08d56fff590b4c9c3e70399c72749f45/state/protocol/protocol_state/state/mutator.go#L47)

In my opinion, everything related to the Epoch should go be governed by one state machine from the perspective of  the `stateMutator`:
* Internally, the epoch state machine would then need to handle the fallback.  So we'll have to do some more work to modularize the logic (including [this code for](https://github.com/onflow/flow-go/blob/ba1f763f08d56fff590b4c9c3e70399c72749f45/state/protocol/protocol_state/state/mutator.go#L205-L258) example), because we shouldn't just error back to the `stateMutator` and re-process all events.  
* I think this logic of mediating between happy path Epoch and fallback should go into the top-level `epoch.stateMachine` 

I think this top-level `epoch.stateMachine` would be the place to document the concurrent processing of the Protocol State Machines. We should emphasize that the `epoch.stateMachine`  is then one of the concurrent state machines running 
Nevertheless, I think the **main place** where we should **document the state machine processing model** is the [`StateMutator` interface](https://github.com/onflow/flow-go/blob/00e4924e0d1b2d69223e5db722b8e6f15ce92399/state/protocol/protocol_state.go#L96-L109) and [`stateMutator` implementation](https://github.com/onflow/flow-go/blob/ba1f763f08d56fff590b4c9c3e70399c72749f45/state/protocol/protocol_state/state/mutator.go#L19-L32).

BUT, lets do that in a follow-up PR. If we agree on this route, may I ask for an issue to be created please 🙏 🙇